src/book.cpp

```diff
@@ -208,22 +227,67 @@ std::string Book::getHumanReadableIdFromPath() const
   return id;
 }
 
+void Book::setFavicon(const std::string& favicon)
+{
+  getMutableDefaultIllustration().data = favicon;
+}
+
+void Book::setFaviconMimeType(const std::string& faviconMimeType)
+{
+  getMutableDefaultIllustration().mimeType = faviconMimeType;
+}
+
 void Book::setPath(const std::string& path)
 {
  m_path = isRelativePath(path)
    ? computeAbsolutePath(getCurrentDirectory(), path)
    : path;
 }
 
-const std::string& Book::getFavicon() const {
-  if (m_favicon.empty() && !m_faviconUrl.empty()) {
+const Book::Illustration& Book::getDefaultIllustration() const
+{
+  for ( const auto& ilPtr : m_illustrations ) {
+    if (ilPtr->width == 48 && ilPtr->height == 48) {
+      return *ilPtr;
+    }
+  }
+  throw NoDefaultIllustration();
+}
+
+Book::Illustration& Book::getMutableDefaultIllustration()
+{
+  try {
+    return const_cast<Illustration&>(getDefaultIllustration());
+  } catch (const NoDefaultIllustration& ) {
+    m_illustrations.push_back(std::make_shared<Illustration>());
+    return *m_illustrations.back();
+  }
+}
+
+const std::string& Book::Illustration::getData() const
+{
+  if (data.empty() && !url.empty()) {
     try {
-      m_favicon = download(m_faviconUrl);
+      data = download(url);
     } catch(...) {
-      std::cerr << "Cannot download favicon from " << m_faviconUrl;
+      std::cerr << "Cannot download favicon from " << url;
     }
   }
-  return m_favicon;
+  return data;
+}
+
+const std::string& Book::getFavicon() const {
+  return getDefaultIllustration().getData();
+}
```

Comment:
As this now throw an exception, it is a change in the API.
At least libxml_dumper expect a empty string if there is no favicon and do not catch exception.
It appears that the java wrapper already catch any exception.

Reply:
> It appears that the java wrapper already catch any exception.

I don't think that is correct. At least [Book's getters](https://github.com/kiwix/libkiwix/blob/9479c0685ddaa71ac5de45b42a962d5b65bfe077/src/wrapper/java/book.cpp#L52-L80) do not handle exceptions.

I added handling of favicon-less books in `LibXMLDumper`, but I am not sure how to proceed with the java wrapper.

Reply:
I reverted the change to `LibXMLDumper`, as well as the earlier change in the API of `Book::getFavicon()`. Now instead of throwing, `Book::getDefaultIllustration()` returns a reference to a static `Illustration` object representing the case of a missing favicon. Thus books without a favicon don't have it included in the result of `getIllustrations()`. Yet the `getFavicon*()` methods keep returning empty strings.

One question is this: if the book has one or more illustrations but none with 48x48 dimensions, maybe calling `getDefaultIllustration()` on it should return one of those existing illustrations rather than the placeholder `missingDefaultIllustration` object?

Reply:
> I don't think that is correct. At least Book's getters do not handle exceptions.

You're right. I was reading [`Reader` code](https://github.com/kiwix/libkiwix/blob/9479c0685ddaa71ac5de45b42a962d5b65bfe077/src/wrapper/java/kiwixreader.cpp#L221-L239)

> I am not sure how to proceed with the java wrapper.

The easier is probably to catch the exception in the wrapper and make it return a empty string if there is no favicon (as before).

> I reverted the change to LibXMLDumper, as well as the earlier change in the API of Book::getFavicon(). Now instead of throwing, Book::getDefaultIllustration() returns a reference to a static Illustration object representing the case of a missing favicon. Thus books without a favicon don't have it included in the result of getIllustrations(). Yet the getFavicon*() methods keep returning empty strings.

I dislike this. The previous version was better. We must design our API following the logic we want to expose and then implement the method with the constraints we have (compatibility and transition are part of the constraints)
The illustration idea is that we can have several illustrations in a zim file (or none). We somehow mandate creator to at least add a 48x48@1 illustration but we cannot force them, and previous zim had no illustration (and sometime no favicon), so we must be prepare to this.

So :
- `Book` have a list of constant illustrations.
- If there is no illustration, this list is empty.
- If a "user" want to get a illustration with a specific size and the illustration is not present, he gets an error (smart choice is a open question. We have this https://github.com/openzim/libzim/issues/541 on libzim side)
- We need to make evolve our data structure (library.xml and opds) to "store" several illustations.

We must handle compatibility with :
- Old zim (this is already handled by libzim itself)
- Old library.xml or opds stream. Here we must read the data and "convert" it to our new scheme.
- Old clients :
  - For cpp clients, we can simply break the API (at least now, because next release will break the API anyway) and ask client to update (and we must do it on kiwix-desktop side)
  - For opds clients, we must ensure that parsing the new opds stream will be ok. As current implementation simply loop on thumbnail and overwrite the favicon data, putting the 48x48@1 image at the end should make things compatible.
  - For xml client it is bit trickier. As the current implementation simply doesn't check the library version (https://github.com/kiwix/libkiwix/blob/master/src/manager.cpp#L58) we must keep a (temporally at least) compatibility by keeping a favicon base64 encoded in the book attributes.
  - For android/java wrapper, we are in the (beginning) middle of moving the wrapper is a separated repository. I suggest we don't change the API here (and so add a compatibility in the wrapper layer). When the wrapper will have moved, we will handle it as a simple cpp client.


Keeping this `getDefaultIllustration` make me think we stick in the middle of a transition where we keep some kind of technical debt.

Reply:
> I don't think that is correct. At least Book's getters do not handle exceptions.

You're right. I was reading [`Reader` code](https://github.com/kiwix/libkiwix/blob/9479c0685ddaa71ac5de45b42a962d5b65bfe077/src/wrapper/java/kiwixreader.cpp#L221-L239)

> I am not sure how to proceed with the java wrapper.

The easier is probably to catch the exception in the wrapper and make it return a empty string if there is no favicon (as before).

> I reverted the change to LibXMLDumper, as well as the earlier change in the API of Book::getFavicon(). Now instead of throwing, Book::getDefaultIllustration() returns a reference to a static Illustration object representing the case of a missing favicon. Thus books without a favicon don't have it included in the result of getIllustrations(). Yet the getFavicon*() methods keep returning empty strings.

I dislike this. The previous version was better. We must design our API following the logic we want to expose and then implement the method with the constraints we have (compatibility and transition are part of the constraints)
The illustration idea is that we can have several illustrations in a zim file (or none). We somehow mandate creator to at least add a 48x48@1 illustration but we cannot force them, and previous zim had no illustration (and sometime no favicon), so we must be prepare to this.

So :
- `Book` have a list of constant illustrations.
- If there is no illustration, this list is empty.
- If a "user" want to get a illustration with a specific size and the illustration is not present, he gets an error (smart choice is a open question. We have this https://github.com/openzim/libzim/issues/541 on libzim side)
- We need to make evolve our data structure (library.xml and opds) to "store" several illustations.

We must handle compatibility with :
- Old zim (this is already handled by libzim itself)
- Old library.xml or opds stream. Here we must read the data and "convert" it to our new scheme.
- Old clients :
  - For cpp clients, we can simply break the API (at least now, because next release will break the API anyway) and ask client to update (and we must do it on kiwix-desktop side)
  - For opds clients, we must ensure that parsing the new opds stream will be ok. As current implementation simply loop on thumbnail and overwrite the favicon data, putting the 48x48@1 image at the end should make things compatible.
  - For xml client it is bit trickier. As the current implementation simply doesn't check the library version (https://github.com/kiwix/libkiwix/blob/master/src/manager.cpp#L58) we must keep a (temporally at least) compatibility by keeping a favicon base64 encoded in the book attributes.
  - For android/java wrapper, we are in the (beginning) middle of moving the wrapper is a separated repository. I suggest we don't change the API here (and so add a compatibility in the wrapper layer). When the wrapper will have moved, we will handle it as a simple cpp client.


Keeping this `getDefaultIllustration` make me think we stick in the middle of a transition where we keep some kind of technical debt.

Reply:
> I dislike this. 

> Keeping this `getDefaultIllustration` make me think we stick in the middle of a transition where we keep some kind of technical debt.

`Book::getDefaultIllustration()` is not part of the public API. It is just an internal function, helping to temporarily preserve the deprecated `getFavicon*()` API. Keeping the `getFavicon*()` API but changing its behavior (throwing an exception when there is no favicon in the ZIM archive) doesn't make any sense - we should either get rid of it completely or ensure perfect backward compatibility. The current state of this PR is along the lines of the latter option.

Reply:
> Keeping the `getFavicon*()` API but changing its behavior (throwing an exception when there is no favicon in the ZIM archive) doesn't make any sense

I'm in favor of removing `getFavicon*()` from the API. Or, if we not removing it, we .keep it as a compatibility layer, using the "correct" API and not changing its behavior.

I'm saying that we must design our API without `getFavicon*()` in mind and then add a compatibility layer.
But if the "normal" API returns a `missingDefaultIllustration` to help implement `getFavicon*()` compatibility functions, there is something wrong.
If there is no illustration, the normal API must return nothing (or empty list, or throw a exception). 

> `Book::getDefaultIllustration()` is not part of the public API

Yes, but it makes me think we are building the implementation around this default illustration. But there is no fundamental default illustration. There are illustrations, with different size, and by default we use to look for the 48x48@1 one. The favicon thing is just a compatibility thing with old behavior.

Reply:
> I'm saying that we must design our API without `getFavicon*()` in mind and then add a compatibility layer.

That's what I've done. The proposed API is the newly introduced public method `Book::getIllustrations()` which returns *only existing illustrations*.  

> > `Book::getDefaultIllustration()` is not part of the public API
> 
> Yes, but it makes me think we are building the implementation around this default illustration.

`Book::getDefaultIllustration()` only serves as the compatibility layer for the deprecated `getFavicon*()` API.

Reply:
> That's what I've done. The proposed API is the newly introduced public method Book::getIllustrations() which returns only existing illustrations.

I've re-read your PR and your right. I've went with a general tirade about public API but we were ok on this. My bad.

> Book::getDefaultIllustration() only serves as the compatibility layer for the deprecated getFavicon*() API

I think this is on the `Book::getMutableDefaultIllustration` we disagree on.
When we parse the xml/opds, we simply have to clear the vector and create a (48x48) Illustration and add it to the vector. Illustration don't have to be mutable (except temporarily at creation).
The `Book::getDefaultIllustration` would simply be a helper to find the 48x48 illustration and throw a exception if not found.

Reply:
> When we parse the xml/opds, we simply have to clear the vector and create a (48x48) Illustration and add it to the vector. Illustration don't have to be mutable (except temporarily at creation).

OK

> The `Book::getDefaultIllustration` would simply be a helper to find the 48x48 illustration and throw a exception if not found.

Do you strongly insist on throwing an exception? What is so bad with the current hack of returning the fake `missingDefaultIllustration` object which helps to avoid the extra `try {} catch {}` blocks in the 3 `getFavicon*()` functions?

Reply:
> Do you strongly insist on throwing an exception?

No. It is not so important. `Book::getDefaultIllustration` returning a fake illustration is ok.